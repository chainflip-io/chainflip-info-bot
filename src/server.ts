import { createBullBoard } from '@bull-board/api';
import { BullMQAdapter } from '@bull-board/api/bullMQAdapter.js';
import { FastifyAdapter } from '@bull-board/fastify';
import { type Queue } from 'bullmq';
import fastify from 'fastify';
import { GraphQLClient } from 'graphql-request';
import env from './env.js';
import logger from './utils/logger.js';

const app = fastify({ loggerInstance: logger, disableRequestLogging: true });

export const explorerClient = new GraphQLClient(env.EXPLORER_GATEWAY_URL);
export const lpClient = new GraphQLClient(env.LP_GATEWAY_URL);

export const createServer = (queues: Queue[]) => {
  const serverAdapter = new FastifyAdapter();

  createBullBoard({
    queues: queues.map((q) => new BullMQAdapter(q)),
    serverAdapter,
  });

  const basePath = '/admin/queues';
  serverAdapter.setBasePath(basePath);
  app.register(serverAdapter.registerPlugin(), { prefix: basePath, basePath: '/' });

  app.get('/health', () => ({ status: 'ok' }));

  app.get('/', (req, res) => res.redirect('/admin/queues', 302));

  return app;
};
