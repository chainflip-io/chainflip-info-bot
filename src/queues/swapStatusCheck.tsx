import React from 'react';
import { formatUsdValue } from '@chainflip/utils/number';
import { type BigNumber } from 'bignumber.js';
import { differenceInMinutes } from 'date-fns';
import { renderToStaticMarkup } from 'react-dom/server';
import { type DispatchJobArgs, type JobConfig, type JobProcessor } from './initialize.js';
import { Bold, ExplorerLink, Line, Trailer } from '../channels/formatting.js';
import { platforms } from '../config.js';
import { humanFriendlyAsset } from '../consts.js';
import env from '../env.js';
import { type ChainflipAsset } from '../graphql/generated/graphql.js';
import getSwapInfo from '../queries/getSwapInfo.js';
import { toFormattedAmount } from '../utils/chainflip.js';
import logger from '../utils/logger.js';

const name = 'swapStatusCheck';
type Name = typeof name;

type Data = {
  swapRequestId: string;
};

type SwapInfo = Awaited<ReturnType<typeof getSwapInfo>>;

declare global {
  interface JobData {
    [name]: Data;
  }
}

const isFresh = (swapInfo: SwapInfo) => {
  const completedAtTimestamp = swapInfo?.completedAt;

  if (!completedAtTimestamp) return false;

  return (
    differenceInMinutes(new Date(), Date.parse(completedAtTimestamp)) <= env.SWAP_MAX_AGE_IN_MINUTES
  );
};

const getSwapStatus = (swapInfo: SwapInfo) => {
  if (swapInfo.completedEventId) return isFresh(swapInfo) ? 'fresh' : 'stale';

  return 'pending';
};

const emoji = (depositValueUsd: string | null | undefined) => {
  if (!depositValueUsd) return null;
  const value = Number.parseFloat(depositValueUsd);
  if (value > 100_000) return 'üê≥';
  if (value > 50_000) return 'ü¶à';
  if (value > 25_000) return 'üêü';
  if (value > 10_000) return 'ü¶Ä';
  return 'ü¶ê';
};

const formatDeltaPrice = (value: string, deltaSign: boolean) => (deltaSign ? `-${value}` : value);

const deltaSign = (delta: number) => {
  if (delta <= -10) return 'üî¥';
  if (delta < -1) return '‚ö™Ô∏è';
  return 'üü¢';
};

const UsdValue = ({ amount }: { amount: BigNumber | null }): React.JSX.Element | null => {
  if (!amount) return null;

  return <> ({formatUsdValue(amount)})</>;
};

const TokenAmount = ({ amount, asset }: { amount: BigNumber; asset: ChainflipAsset }) => (
  <>
    {toFormattedAmount(amount)} {humanFriendlyAsset[asset]}
  </>
);

const buildMessageData = ({
  swapInfo,
}: {
  swapInfo: SwapInfo;
}): Extract<DispatchJobArgs, { name: 'messageRouter' }>[] =>
  platforms.map((platform) => {
    const message = renderToStaticMarkup(
      <>
        <Line>
          {emoji(swapInfo.originalDepositValueUsd)} Swap{' '}
          <Bold platform={platform}>
            <ExplorerLink platform={platform} path={`swaps/${swapInfo.requestId}`} prefer="link">
              #{swapInfo.requestId}
            </ExplorerLink>
          </Bold>
        </Line>
        <Line>
          üì•{' '}
          <Bold platform={platform}>
            <TokenAmount amount={swapInfo.depositAmount} asset={swapInfo.sourceAsset} />
          </Bold>
          <UsdValue amount={swapInfo.depositValueUsd} />
        </Line>
        {swapInfo.egressAmount && (
          <Line>
            üì§{' '}
            <Bold platform={platform}>
              <TokenAmount amount={swapInfo.egressAmount} asset={swapInfo.destinationAsset} />
            </Bold>
            <UsdValue amount={swapInfo.egressValueUsd} />
          </Line>
        )}
        {swapInfo.refundAmount && (
          <Line>
            ‚Ü©Ô∏è{' '}
            <Bold platform={platform}>
              <TokenAmount amount={swapInfo.refundAmount} asset={swapInfo.sourceAsset} />
            </Bold>
            <UsdValue amount={swapInfo.refundValueUsd} />
          </Line>
        )}
        {swapInfo.duration && (
          <Line>
            ‚è±Ô∏è Took: <Bold platform={platform}>{swapInfo.duration}</Bold>
          </Line>
        )}
        {swapInfo.priceDelta !== null && swapInfo.priceDeltaPercentage && (
          <Line>
            {deltaSign(Number(swapInfo.priceDeltaPercentage))} Delta:{' '}
            <Bold platform={platform}>
              {formatDeltaPrice(
                formatUsdValue(Math.abs(swapInfo.priceDelta)),
                Number(swapInfo.priceDeltaPercentage) < 0,
              )}
            </Bold>{' '}
            ({swapInfo.priceDeltaPercentage}%)
          </Line>
        )}
        {swapInfo.dcaChunks && (
          <Line>
            üìì Chunks: <Bold platform={platform}>{swapInfo.dcaChunks}</Bold>
          </Line>
        )}
        {swapInfo.boostFee.gt(0) && (
          <Line>
            ‚ö° <Bold platform={platform}>Boosted</Bold> for{' '}
            <Bold platform={platform}>{formatUsdValue(swapInfo.boostFee)}</Bold>
          </Line>
        )}
        {swapInfo.brokerIdAndAlias && (
          <Line>
            üè¶ via{' '}
            <Bold platform={platform}>
              <ExplorerLink
                platform={platform}
                path={`/brokers/${swapInfo.brokerIdAndAlias.brokerId}`}
                prefer="text"
              >
                {swapInfo.brokerIdAndAlias.alias}
              </ExplorerLink>
            </Bold>
          </Line>
        )}
        {swapInfo.affiliatesIdsAndAliases?.length ? (
          <Line>
            üè∞ Affiliate:{' '}
            {swapInfo.affiliatesIdsAndAliases?.map((affiliate) => (
              <Bold key={affiliate.brokerId} platform={platform}>
                <ExplorerLink
                  platform={platform}
                  path={`/brokers/${affiliate.brokerId}`}
                  prefer="text"
                >
                  {affiliate.alias}
                </ExplorerLink>{' '}
              </Bold>
            ))}
          </Line>
        ) : null}
        <Trailer platform={platform} />
      </>,
    ).trimEnd();

    return {
      name: 'messageRouter' as const,
      data: {
        platform,
        message,
        filterData: { name: 'NEW_SWAP', usdValue: Number(swapInfo?.egressValueUsd || 0) },
      },
    };
  });

const processJob: JobProcessor<Name> = (dispatchJobs) => async (job) => {
  logger.info(`Checking swap #${job.data.swapRequestId}`);
  const swapInfo = await getSwapInfo(job.data.swapRequestId);

  const jobs = [] as DispatchJobArgs[];

  if (
    swapInfo.completedEventId &&
    (swapInfo.egressAmount === null || swapInfo.egressAmount.isZero())
  ) {
    logger.info(`Swap egress amount is zero, so it was refunded`);
    return;
  }

  const status = getSwapStatus(swapInfo);

  switch (status) {
    case 'fresh':
      jobs.push(...buildMessageData({ swapInfo }));
      logger.info(`Swap #${swapInfo.requestId} is fresh, job was added in a queue`);
      break;

    case 'stale':
      logger.warn(`Swap #${swapInfo.requestId} is stale`);
      break;

    case 'pending':
      jobs.push({
        name: 'scheduler',
        data: [{ name, data: { swapRequestId: job.data.swapRequestId } }],
        opts: { delay: 10_000 },
      } as const);
      logger.info(`Swap #${swapInfo.requestId} is not completed, pushed to a scheduler`);
      break;
  }

  await dispatchJobs(jobs);
};

export const config: JobConfig<typeof name> = {
  name,
  processJob,
};
