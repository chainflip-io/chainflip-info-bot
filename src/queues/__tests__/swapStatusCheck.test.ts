import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import pendingSwapStats from './pendingSwapStats.json' with { type: 'json' };
import swapInfoStats from '../../queries/__tests__/swapInfo.json' with { type: 'json' };

import { explorerClient } from '../../server.js';
import { config } from '../swapStatusCheck.js';

beforeEach(() => {
  vi.useFakeTimers();
});

afterEach(() => {
  vi.useRealTimers();
});

describe('swapStatusCheck', () => {
  it('check fresh swap status and send swap info message', async () => {
    vi.setSystemTime(new Date('2024-10-25T12:42:30+00:00'));
    vi.mocked(explorerClient.request).mockResolvedValue(swapInfoStats);

    const dispatchJobs = vi.fn();

    await config.processJob(dispatchJobs)({
      data: {
        swapRequestId: '77697',
      } as JobData['swapStatusCheck'],
    } as any);

    expect(dispatchJobs.mock.lastCall).toMatchSnapshot();
  });

  it('check pending swap status and move to a scheduler', async () => {
    vi.setSystemTime(new Date('2024-10-25T12:42:30+00:00'));
    vi.mocked(explorerClient.request).mockResolvedValue(pendingSwapStats);

    const dispatchJobs = vi.fn();
    await config.processJob(dispatchJobs)({
      data: {
        swapRequestId: '77697',
      } as JobData['swapStatusCheck'],
    } as any);

    expect(dispatchJobs.mock.lastCall).toMatchInlineSnapshot(`
      [
        [
          {
            "data": [
              {
                "data": {
                  "swapRequestId": "77697",
                },
                "name": "swapStatusCheck",
                "opts": {
                  "delay": 10000,
                },
              },
            ],
            "name": "scheduler",
          },
        ],
      ]
    `);
  });

  it('check stale swap status', async () => {
    vi.mocked(explorerClient.request).mockResolvedValue(swapInfoStats);

    const dispatchJobs = vi.fn();
    await config.processJob(dispatchJobs)({
      data: {
        swapRequestId: '77697',
      } as JobData['swapStatusCheck'],
    } as any);

    expect(dispatchJobs.mock.lastCall).toMatchInlineSnapshot(`
      [
        [],
      ]
    `);
  });
});
