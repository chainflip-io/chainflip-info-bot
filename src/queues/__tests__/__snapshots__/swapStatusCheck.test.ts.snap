// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`swapStatusCheck > check fresh swap status and send swap info message 1`] = `
[
  [
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 5611.3799573378,
        },
        "message": "🦐 Swap <strong><a href="https://scan.chainflip.io/swaps/77697">#77697</a></strong>
📥 <strong>5,000 FLIP on Ethereum</strong> ($5,568.07)
📤 <strong>5,616.094932 USDT on Ethereum</strong> ($5,611.38)
⏱️ Took: <strong>1 min</strong>
🟢 Delta: <strong>$43.31</strong> (0.77%)
📓 Chunks: <strong>2/2</strong>
🏦 via <strong><a href="https://scan.chainflip.io/brokers/cFLRQDfEdmnv6d2XfHJNRBQHi4fruPMReLSfvB8WWD2ENbqj7">Chainflip Swapping</a></strong>",
        "platform": "telegram",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 5611.3799573378,
        },
        "message": "🦐 Swap **[#77697](https://scan.chainflip.io/swaps/77697)**
📥 **5,000 FLIP on Ethereum** ($5,568.07)
📤 **5,616.094932 USDT on Ethereum** ($5,611.38)
⏱️ Took: **1 min**
🟢 Delta: **$43.31** (0.77%)
📓 Chunks: **2/2**
🏦 via **[Chainflip Swapping](https://scan.chainflip.io/brokers/cFLRQDfEdmnv6d2XfHJNRBQHi4fruPMReLSfvB8WWD2ENbqj7)**",
        "platform": "discord",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 5611.3799573378,
        },
        "message": "🦐 Swap https://scan.chainflip.io/swaps/77697
📥 5,000 FLIP on Ethereum ($5,568.07)
📤 5,616.094932 USDT on Ethereum ($5,611.38)
⏱️ Took: 1 min
🟢 Delta: $43.31 (0.77%)
📓 Chunks: 2/2
🏦 via Chainflip Swapping
#chainflip $flip",
        "platform": "twitter",
      },
      "name": "messageRouter",
    },
  ],
]
`;

exports[`swapStatusCheck > subtracts out the refund egress amount from the ingress amount 1`] = `
[
  [
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 17206.1360813008,
        },
        "message": "🐟 Swap <strong><a href="https://scan.chainflip.io/swaps/103045">#103045</a></strong>
📥 <strong>13,348.70221 FLIP on Ethereum</strong> ($17,038.94)
📤 <strong>5.939236 ETH on Ethereum</strong> ($17,206.14)
↩️ <strong>24,455.099318 FLIP on Ethereum</strong> ($31,269.46)
⏱️ Took: <strong>18 min</strong>
🟢 Delta: <strong>$167.20</strong> (0.98%)
📓 Chunks: <strong>6/17</strong>
🏦 via <strong><a href="https://scan.chainflip.io/brokers/cFLRQDfEdmnv6d2XfHJNRBQHi4fruPMReLSfvB8WWD2ENbqj7">Chainflip Swapping</a></strong>",
        "platform": "telegram",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 17206.1360813008,
        },
        "message": "🐟 Swap **[#103045](https://scan.chainflip.io/swaps/103045)**
📥 **13,348.70221 FLIP on Ethereum** ($17,038.94)
📤 **5.939236 ETH on Ethereum** ($17,206.14)
↩️ **24,455.099318 FLIP on Ethereum** ($31,269.46)
⏱️ Took: **18 min**
🟢 Delta: **$167.20** (0.98%)
📓 Chunks: **6/17**
🏦 via **[Chainflip Swapping](https://scan.chainflip.io/brokers/cFLRQDfEdmnv6d2XfHJNRBQHi4fruPMReLSfvB8WWD2ENbqj7)**",
        "platform": "discord",
      },
      "name": "messageRouter",
    },
    {
      "data": {
        "filterData": {
          "name": "NEW_SWAP",
          "usdValue": 17206.1360813008,
        },
        "message": "🐟 Swap https://scan.chainflip.io/swaps/103045
📥 13,348.70221 FLIP on Ethereum ($17,038.94)
📤 5.939236 ETH on Ethereum ($17,206.14)
↩️ 24,455.099318 FLIP on Ethereum ($31,269.46)
⏱️ Took: 18 min
🟢 Delta: $167.20 (0.98%)
📓 Chunks: 6/17
🏦 via Chainflip Swapping
#chainflip $flip",
        "platform": "twitter",
      },
      "name": "messageRouter",
    },
  ],
]
`;
